src/diffusers/quantizers/torchao/torchao_quantizer.py

```diff
@@ -0,0 +1,253 @@
+                logger.warning(
+                    f"You are trying to set torch_dtype to {torch_dtype} for int4/int8/uintx quantization, but "
+                    f"only bfloat16 is supported right now. Please set `torch_dtype=torch.bfloat16`."
+                )
+
+        if torch_dtype is None:
+            # we need to set the torch_dtype, otherwise we have dtype mismatch when performing the quantized linear op
+            logger.warning(
+                "Overriding `torch_dtype` with `torch_dtype=torch.bfloat16` due to requirements of `torchao` "
+                "to enable model loading in different precisions. Pass your own `torch_dtype` to specify the "
+                "dtype of the remaining non-linear layers, or pass torch_dtype=torch.bfloat16, to remove this warning."
+            )
+            torch_dtype = torch.bfloat16
+
+        return torch_dtype
+
+    def adjust_target_dtype(self, target_dtype: "torch.dtype") -> "torch.dtype":
+        supported_dtypes = (
+            torch.int8,
+            torch.float8_e4m3fn,
+            torch.float8_e5m2,
+            torch.uint1,
+            torch.uint2,
+            torch.uint3,
+            torch.uint4,
+            torch.uint5,
+            torch.uint6,
+            torch.uint7,
+        )
+        if isinstance(target_dtype, supported_dtypes):
+            return target_dtype
+
+        raise ValueError(
+            f"You are using `device_map='auto'` on a TorchAO quantized model but a suitable target dtype "
+            f"could not be inferred. The supported target_dtypes are: {supported_dtypes}. If you think the "
+            f"dtype you are using should be supported, please open an issue at https://github.com/huggingface/diffusers/issues."
+        )
+
+    def adjust_max_memory(self, max_memory: Dict[str, Union[int, str]]) -> Dict[str, Union[int, str]]:
+        # TODO(aryan): Not sure what to do here @sayakpaul, @DN6
+        # need more space for the quantization parameters (e.g. scale). Tested with int4 wo and group size = 128
+        max_memory = {key: val * 0.9 for key, val in max_memory.items()}
+        return max_memory
+
+    def check_if_quantized_param(
+        self,
+        model: "ModelMixin",
+        param_value: "torch.Tensor",
+        param_name: str,
+        state_dict: Dict[str, Any],
+        **kwargs,
+    ) -> bool:
+        param_device = kwargs.pop("param_device", None)
+        # check if the param_name is not in self.modules_to_not_convert
+        if any((key + "." in param_name) or (key == param_name) for key in self.modules_to_not_convert):
+            return False
+        elif param_device == "cpu" and self.offload:
+            # We don't quantize weights that we offload
+            return False
+        else:
+            # we only quantize the weight of nn.Linear
+            module, tensor_name = get_module_from_name(model, param_name)
+            return isinstance(module, torch.nn.Linear) and (tensor_name == "weight")
```

Comment:
@jerryzh168 are there any better ways to detect if a module has quantization from torchao applied? Checking for just `nn.Linear` and its name feel a bit thin.

Reply:
do we need to do this checking? one difference between torchao and bitsandbytes (or all other quant libraries probably) is that torchao is using tensor subclass, so we can just load the quantized state dict into original floating point model without changes/or checks and the model will be turned into a quantized model.

also `check_if_quantized_param` is this to check the params **to quantize** or **quantized** params?

Reply:
We check it to do a further inspection on the shape as well as to handle any parameter creation. But I guess it's fine with `torchao` because of the reasons you mentioned. @SunMarc WDYT?

Reply:
This check is here to check the params to quantize and quantized params. We can skip these here but eventually, we will need to do these checks in `create_quantized_params`. 

> so we can just load the quantized state dict into original floating point model without changes

Nice to know that. Right now, when using quantization, we need to iterate through each params but we can think a new path where we just do model.load_state_dict().